### ISSUE TITLE
AI老仙，法力无边。

### ISSUE THREAD

**Issue by shuidongliu-shui** [AngusMonroe/ContrastPool#9]
在AI老仙的大力支持下，总算是鼓捣出了生成bin文件的代码，但是它和BrainNet.py有各种各样的兼容问题，我等等明天接着搞吧，先把这个代码放这。主要我还是win系统，我修改baseline脚本还不知道为啥能运行但是有那种奇奇怪怪的中文。等明天吧，我觉得应该能把这个复现出来。
import os
import numpy as np
import scipy.io
import networkx as nx
import dgl
import torch
import glob
import re
import json
from tqdm import tqdm
from dgl.data.utils import save_graphs

def load_and_validate_matrices(file_list):
    """加载并验证MAT文件，确保所有矩阵为100x100"""
    valid_matrices = []
    valid_files = []
    error_log = []

    for file_path in tqdm(file_list, desc="验证矩阵文件"):
        try:
            # 加载MAT文件
            mat_data = scipy.io.loadmat(file_path)
            matrix = mat_data['data']

            # 维度校验与修正
            if matrix.ndim == 3 and matrix.shape[:2] == (100, 100):
                matrix = matrix.reshape(100, 100)
            elif matrix.shape != (100, 100):
                raise ValueError(f"无效维度 {matrix.shape}")

            # 标准化处理
            np.fill_diagonal(matrix, 0)  # 清除自连接
            matrix[matrix < 0] = 0      # 去除负相关

            valid_matrices.append(matrix.astype(np.float32))  # 节省内存
            valid_files.append(file_path)

        except Exception as e:
            error_log.append(f"{os.path.basename(file_path)}: {str(e)}")

    # 打印错误报告
    if error_log:
        print(f"\n发现 {len(error_log)} 个错误文件:")
        print("\n".join(error_log[:3]))

    return np.array(valid_matrices), valid_files

def build_dgl_graph(matrix, label):
    """将相关矩阵转换为带权DGL图（关键修改点）"""
    # 创建网络图（带权有向图）
    src_nodes, dst_nodes = np.where(matrix > 0)
    edge_weights = matrix[src_nodes, dst_nodes]

    # 构建DGL图
    dgl_g = dgl.graph((src_nodes, dst_nodes))
    dgl_g.edata['weight'] = torch.FloatTensor(edge_weights)

    # 修改节点特征键名（'feat' -> 'N_features'）
    dgl_g.ndata['N_features'] = torch.eye(100)  # 关键修改点
    dgl_g.graph_label = torch.LongTensor([label])

    return dgl_g

def construct_abide_dataset():
    # 路径配置
    BASE_DIR = r"C:\ABIDE"
    OUTPUT_DIR = os.path.join(BASE_DIR, "processed_dataset")
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    # 文件查找模式
    file_pattern = os.path.join(BASE_DIR, "**", "sub-*_schaefer100_correlation_matrix.mat")
    mat_files = glob.glob(file_pattern, recursive=True)

    if not mat_files:
        raise FileNotFoundError(f"未找到匹配文件: {file_pattern}")

    # 数据加载验证
    matrices, valid_files = load_and_validate_matrices(mat_files)
    print(f"\n✅ 成功加载 {len(matrices)} 个有效矩阵")

    # 创建组别编码
    group_map = {}
    graph_list = []
    label_list = []

    # 处理每个有效文件
    for file_path, matrix in zip(tqdm(valid_files, desc="构建图数据"), matrices):
        # 从文件名提取组别（示例：sub-autism12345... 或 sub-control67890...）
        match = re.search(r"sub-([a-zA-Z]+)\d+_", os.path.basename(file_path))
        if not match:
            continue

        group = match.group(1).lower()
        if group not in group_map:
            group_map[group] = len(group_map)

        # 构建DGL图
        dgl_graph = build_dgl_graph(matrix, group_map[group])
        graph_list.append(dgl_graph)
        label_list.append(group_map[group])

    # 保存数据集（使用指定名称）
    output_path = os.path.join(OUTPUT_DIR, "abide_schaefer100.bin")
    save_graphs(output_path, graph_list, {"labels": torch.LongTensor(label_list)})

    # 保存组别映射
    with open(os.path.join(OUTPUT_DIR, "group_mapping.json"), "w") as f:
        json.dump(group_map, f, indent=2)

    print(f"\n🔥 数据集构建完成！")
    print(f"📁 输出文件: {output_path}")
    print(f"📊 类别分布: { {k: label_list.count(v) for k, v in group_map.items()} }")

if __name__ == "__main__":
    construct_abide_dataset()
